resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/hadoopsteps/HadoopKerberosKeytabResolverStep.scala

```diff
@@ -0,0 +1,171 @@
+        renewer = jobUserUGI.getShortUserName
+        logDebug(s"Renewer is: $renewer")
+        credentials = new Credentials(originalCredentials)
+        dfs.addDelegationTokens(renewer, credentials)
+        // This is difficult to Mock and will require refactoring
+        tokens = credentials.getAllTokens.asScala
+        logDebug(s"Tokens: ${credentials.toString}")
+        logDebug(s"All tokens: ${tokens.mkString(",")}")
+        logDebug(s"All secret keys: ${credentials.getAllSecretKeys}")
+        null
+      }})
+    credentials.getAllTokens.asScala.isEmpty
+    tokens.isEmpty
+    if (tokens.isEmpty) logError("Did not obtain any Delegation Tokens")
+    val data = serialize(credentials)
+    val renewalTime = getTokenRenewalInterval(tokens, hadoopConf).getOrElse(Long.MaxValue)
+    val currentTime: Long = System.currentTimeMillis()
+    val initialTokenDataKeyName = s"$KERBEROS_SECRET_LABEL_PREFIX-$currentTime-$renewalTime"
+    val secretDT =
+      new SecretBuilder()
+        .withNewMetadata()
+          .withName(HADOOP_KERBEROS_SECRET_NAME)
+          .withLabels(Map("refresh-hadoop-tokens" -> "yes").asJava)
+          .endMetadata()
+          .addToData(initialTokenDataKeyName, Base64.encodeBase64String(data))
+      .build()
+    val bootstrapKerberos = new KerberosTokenConfBootstrapImpl(
+      HADOOP_KERBEROS_SECRET_NAME,
+      initialTokenDataKeyName,
+      jobUserUGI.getShortUserName)
+    val withKerberosEnvPod = bootstrapKerberos.bootstrapMainContainerAndVolumes(
+      PodWithMainContainer(
+        hadoopConfigSpec.driverPod,
+        hadoopConfigSpec.driverContainer))
+    hadoopConfigSpec.copy(
+      additionalDriverSparkConf =
+        hadoopConfigSpec.additionalDriverSparkConf ++ Map(
+          HADOOP_KERBEROS_CONF_ITEM_KEY -> initialTokenDataKeyName,
+          HADOOP_KERBEROS_CONF_SECRET -> HADOOP_KERBEROS_SECRET_NAME),
+      driverPod = withKerberosEnvPod.pod,
+      driverContainer = withKerberosEnvPod.mainContainer,
+      dtSecret = Some(secretDT),
+      dtSecretName = HADOOP_KERBEROS_SECRET_NAME,
+      dtSecretItemKey = initialTokenDataKeyName)
+  }
+
+  // Functions that should be in Core with Rebase to 2.3
+  @deprecated("Moved to core in 2.2", "2.2")
+  private def getTokenRenewalInterval(
+    renewedTokens: Iterable[Token[_ <: TokenIdentifier]],
+    hadoopConf: Configuration): Option[Long] = {
+      val renewIntervals = renewedTokens.filter {
+        _.decodeIdentifier().isInstanceOf[AbstractDelegationTokenIdentifier]}
+        .flatMap { token =>
+        Try {
+          val newExpiration = token.renew(hadoopConf)
+          val identifier = token.decodeIdentifier().asInstanceOf[AbstractDelegationTokenIdentifier]
+          val interval = newExpiration - identifier.getIssueDate
+          logInfo(s"Renewal interval is $interval for token ${token.getKind.toString}")
+          interval
+        }.toOption}
+      if (renewIntervals.isEmpty) None else Some(renewIntervals.min)
```

Comment:
Just as a note to myself, I see now that we'll return the earliest expiration time in case there are multiple tokens.

Reply:
Correct

Reply:
Maybe we should add a comment about this, because it wasn't easy to discover.